### PR TITLE
fix: use FileCopier for copying files and queue creation of symlinks

### PR DESCRIPTION
**Pull request by mmaietta**
Fixes: https://github.com/electron-userland/electron-builder/issues/8655

Adds HelloWorld.framework test case (along with build script) to test folder structures containing a plethora of symlinks.
Test case tree structure is below:
```
/private/var/folders/y3/r1xf5xzn5pddms2gslqh14sh0000gn/T/et-5a873b0a5efb7affcd771efe4ab31bfa/t-uhAANC/test-project-0/dist/linux-arm64-unpacked/resources/app
├── foo.js -> index.js
├── index.js
├── node_modules
│   ├── debug
│   │   ├── LICENSE
│   │   ├── dist
│   │   │   └── debug.js
│   │   ├── package.json
│   │   └── src
│   │       ├── browser.js
│   │       ├── common.js
│   │       ├── index.js
│   │       └── node.js
│   ├── hello-world
│   │   ├── hello-world-framework.sh
│   │   ├── lib
│   │   │   └── Release
│   │   │       └── Hello.framework
│   │   │           ├── Headers -> Versions/Current/Headers
│   │   │           ├── Hello -> Versions/Current/Hello
│   │   │           ├── Info.plist -> Versions/Current/Info.plist
│   │   │           ├── Modules -> Versions/Current/Modules
│   │   │           └── Versions
│   │   │               ├── A
│   │   │               │   ├── Headers
│   │   │               │   │   └── HelloFramework.h
│   │   │               │   ├── Hello
│   │   │               │   ├── Info.plist
│   │   │               │   └── Modules
│   │   │               │       ├── Hello.abi.json
│   │   │               │       ├── Hello.swiftdoc
│   │   │               │       ├── Hello.swiftmodule
│   │   │               │       │   ├── arm64.private.swiftinterface
│   │   │               │       │   ├── arm64.swiftinterface
│   │   │               │       │   ├── arm64.swiftmodule
│   │   │               │       │   ├── x86_64.private.swiftinterface
│   │   │               │       │   ├── x86_64.swiftinterface
│   │   │               │       │   └── x86_64.swiftmodule
│   │   │               │       ├── Hello.swiftsourceinfo
│   │   │               │       └── module.modulemap
│   │   │               └── Current -> A
│   │   └── package.json
│   └── ms
│       ├── index.js
│       ├── license.md
│       └── package.json
└── package.json
```

### CODE REVIEW

File: packages/app-builder-lib/src/asar/asarUtil.ts [hunk 1]
```diff
@@ -69,95 +77,104 @@ export class AsarPackager {
...
   }
 
   private async detectAndCopy(fileSets: ResolvedFileSet[], cancellationToken: CancellationToken) {
     const taskManager = new AsyncTaskManager(cancellationToken)
     const unpackedPaths = new Set<string>()
     const copiedFiles = new Set<string>()
 
+    const createdSourceDirs = new Set<string>()
+    const links: Array<Link> = []
+    const symlinkType = platform() === "win32" ? "junction" : "file"
```
Comment (beyondkmp) [inline packages/app-builder-lib/src/asar/asarUtil.ts:94]: If the symlink points to a directory, would this cause any issues?

File: packages/app-builder-lib/src/asar/asarUtil.ts [hunk 2]
```diff
@@ -69,95 +77,104 @@ export class AsarPackager {
...
-        destination,
-        stat,
-        fileSet: { src: sourceDir },
-      } = options
-      copiedFiles.add(destination)
-
-      // If transformed data, skip symlink logic
-      if (transformedData) {
-        return this.copyFileOrData(transformedData, source, destination, stat)
+    const writeFileOrQueueSymlink = async (options: { transformedData: string | Buffer | undefined; file: string; destFile: string; stat: fs.Stats; fileSet: ResolvedFileSet }) => {
```
Comment (beyondkmp) [inline packages/app-builder-lib/src/asar/asarUtil.ts:103]: How about `writeFileOrProcessSymlinkConcurrently `?

File: packages/app-builder-lib/src/asar/asarUtil.ts (hunk 1 above)
Comment (mmaietta) [reply packages/app-builder-lib/src/asar/asarUtil.ts:94]: This is the test case folder tree. Several symlinks are directories.
```
/private/var/folders/y3/r1xf5xzn5pddms2gslqh14sh0000gn/T/et-5a873b0a5efb7affcd771efe4ab31bfa/t-uhAANC/test-project-0/dist/linux-arm64-unpacked/resources/app
├── foo.js -> index.js
├── index.js
├── node_modules
│   ├── debug
│   │   ├── LICENSE
│   │   ├── dist
│   │   │   └── debug.js
│   │   ├── package.json
│   │   └── src
│   │       ├── browser.js
│   │       ├── common.js
│   │       ├── index.js
│   │       └── node.js
│   ├── hello-world
│   │   ├── hello-world-framework.sh
│   │   ├── lib
│   │   │   └── Release
│   │   │       └── Hello.framework
│   │   │           ├── Headers -> Versions/Current/Headers
│   │   │           ├── Hello -> Versions/Current/Hello
│   │   │           ├── Info.plist -> Versions/Current/Info.plist
│   │   │           ├── Modules -> Versions/Current/Modules
│   │   │           └── Versions
│   │   │               ├── A
│   │   │               │   ├── Headers
│   │   │               │   │   └── HelloFramework.h
│   │   │               │   ├── Hello
│   │   │               │   ├── Info.plist
│   │   │               │   └── Modules
│   │   │               │       ├── Hello.abi.json
│   │   │               │       ├── Hello.swiftdoc
│   │   │               │       ├── Hello.swiftmodule
│   │   │               │       │   ├── arm64.private.swiftinterface
│   │   │               │       │   ├── arm64.swiftinterface
│   │   │               │       │   ├── arm64.swiftmodule
│   │   │               │       │   ├── x86_64.private.swiftinterface
│   │   │               │       │   ├── x86_64.swiftinterface
│   │   │               │       │   └── x86_64.swiftmodule
│   │   │               │       ├── Hello.swiftsourceinfo
│   │   │               │       └── module.modulemap
│   │   │               └── Current -> A
│   │   └── package.json
│   └── ms
│       ├── index.js
│       ├── license.md
│       └── package.json
└── package.json
```

Note, I tested this on mac and linux, I'm not sure on windows actually, but I was mirroring this logic here https://github.com/electron-userland/electron-builder/blob/df07101305b24a62731f6b4359228c508c5c76c7/packages/builder-util/src/fs.ts#L298

**Review by beyondkmp — DISMISSED** [review]

**Comment by beyondkmp** [comment]
There's another scenario where a symlink points to an empty directory. The actual empty directory probably won't be copied, and then the symlink will throw an error.

In this scenario, should we just delete this kind of symlink without any further processing?

**Comment by mmaietta** [comment]
Great callout @beyondkmp , looks like that issue is pre-existing elsewhere in the codebase
I added a `Resources` symlink to an empty directory and received this error. That's a separate issue from the `electron/asar` changes AFAICT though since it's before the asar packaging step
```
    ENOENT: no such file or directory, stat '/tmp/et-db411cd21f3bbd882f3a5a30c58db6ff/t-dVdcd3/test-project-0/hello-world/lib/Release/Hello.framework/Resources'

      169 |
      170 |     for (const [file, index] of symlinkFiles) {
    > 171 |       const resolvedPath = realpathSync(file)
          |                                        ^
      172 |       if (emptyDirs.has(resolvedPath)) {
      173 |         // delete symlink file if target is a empty dir
      174 |         result[index] = undefined

      at NodeModuleCopyHelper.collectNodeModules (../packages/app-builder-lib/src/util/NodeModuleCopyHelper.ts:171:40)
      at collectNodeModules (../packages/app-builder-lib/src/util/appFileCopier.ts:195:19)
      at computeNodeModuleFileSets (../packages/app-builder-lib/src/util/appFileCopier.ts:207:5)
```

The folder tree:
```
test/fixtures/test-app-symlink-framework/hello-world/lib/Release
└── Hello.framework
    ├── Headers -> Versions/Current/Headers
    ├── Hello -> Versions/Current/Hello
    ├── Info.plist -> Versions/Current/Info.plist
    ├── Modules -> Versions/Current/Modules
    ├── Resources -> Versions/Current/Resources
    └── Versions
        ├── A
        │   ├── Headers
        │   │   └── HelloFramework.h
        │   ├── Hello
        │   ├── Info.plist
        │   ├── Modules
        │   │   ├── Hello.abi.json
        │   │   ├── Hello.swiftdoc
        │   │   ├── Hello.swiftmodule
        │   │   │   ├── arm64.private.swiftinterface
        │   │   │   ├── arm64.swiftinterface
        │   │   │   ├── arm64.swiftmodule
        │   │   │   ├── x86_64.private.swiftinterface
        │   │   │   ├── x86_64.swiftinterface
        │   │   │   └── x86_64.swiftmodule
        │   │   ├── Hello.swiftsourceinfo
        │   │   └── module.modulemap
        │   └── Resources
        └── Current -> A

12 directories, 15 files
```

**Pull request merged**